### ISSUE TITLE
TensorRT virtualMemoryBuffer internal error

### ISSUE THREAD

**Issue by frank-wei** [pytorch/torchdynamo#111]
Observed the error msg after run. Maybe related with how torchdynamo release the resources? 

python torchbench.py -dcuda --speedup-fx2trt-fp16 --only mobilenet_v2 
cuda eval  mobilenet_v2                       [04/05/2022-15:43:11] [TRT] [W] TensorRT was linked against cuBLAS/cuBLAS LT 11.6.3 but loaded cuBLAS/cuBLAS LT 11.5.1
[04/05/2022-15:43:14] [TRT] [W] TensorRT was linked against cuDNN 8.2.1 but loaded cuDNN 8.0.5
[04/05/2022-15:43:48] [TRT] [W] TensorRT was linked against cuBLAS/cuBLAS LT 11.6.3 but loaded cuBLAS/cuBLAS LT 11.5.1
[04/05/2022-15:43:48] [TRT] [W] TensorRT was linked against cuDNN 8.2.1 but loaded cuDNN 8.0.5
[04/05/2022-15:43:48] [TRT] [W] TensorRT was linked against cuBLAS/cuBLAS LT 11.6.3 but loaded cuBLAS/cuBLAS LT 11.5.1
[04/05/2022-15:43:48] [TRT] [W] TensorRT was linked against cuDNN 8.2.1 but loaded cuDNN 8.0.5

Supported node types in the model:
acc_ops.conv2d: ((), {'input': torch.float16, 'weight': torch.float16})
acc_ops.batch_norm: ((), {'input': torch.float16, 'running_mean': torch.float16, 'running_var': torch.float16, 'weight': torch.float16, 'bias': torch.float16})
acc_ops.hardtanh: ((), {'input': torch.float16})
acc_ops.add: ((), {'input': torch.float16, 'other': torch.float16})
acc_ops.adaptive_avg_pool2d: ((), {'input': torch.float16})
acc_ops.flatten: ((), {'input': torch.float16})
acc_ops.linear: ((), {'input': torch.float16, 'weight': torch.float16, 'bias': torch.float16})

Unsupported node types in the model:

graph is split into _run_on_acc_0
Similarity score=0.9999595284461975
7.243x p=0.00
**Unexpected Internal Error: [virtualMemoryBuffer.cpp::~StdVirtualMemoryBufferImpl::121] Error Code 1: Cuda Runtime (driver shutting down)**

**Comment by jansel** [comment]
I believe `virtualMemoryBuffer.cpp` is part of TensorRT based on quick Google search results.  So this sounds like some internal TensorRT error.  Any reason to think this is TorchDynamo related?

**Comment by jlucier** [comment]
@frank-wei @jansel any updates on this? I think I am experiencing this issue on 8.0.1, but the error output does not include as much information. I see this exact message on 8.2.4. Notably, my code has not changed and everything worked fine with 7.2.3.

This only occurs when I use a TensorRT engine in the same process as a GUI application when the process is terminating. I've detailed this more here as well: https://forums.developer.nvidia.com/t/tensorrt-8-nvinfer1-cudaruntimeerror-gui-application/188887

**Comment by jansel** [comment]
I'll let @frank-wei answer as I think it is related to TRT.

**Comment by frank-wei** [comment]
@jlucier , I am not sure the root cause yet since the log is not informative. For torchdynamo+fx2trt, it generates this error msg in torchbench.py but not in other places like some local test I made. My plan is to simplify the code and carefully remove some unnecessary and resource collection code and check if the problem persists. 
I would suggest you the same idea and you may find the root cause later. NV's software is flaky in some cases.

**Comment by jlucier** [comment]
@frank-wei Sounds good! Thanks!

You don't have to tell me haha! I'm very well aware that NV's software is bit brutal. I've been using their Jetson platform, CUDA, and TensorRT since 2018 and I've lost countless hours to stuff like this :laughing:

**Comment by jlucier** [comment]
@frank-wei I actually think I figured my issue out...

I use the TensorRT C++ API directly and expose an interface for performing inference to Python (using pybind11). What I noticed is for some reason the destructor (in C++) for my object which holds the TRT engine and CUDA stream was not getting called when the program exited. As a result, the TRT engine and other objects were not explicitly being destroyed (with `delete`) which I guess was making TRT and CUDA unhappy.

I used [atexit](https://docs.python.org/3/library/atexit.html) in Python to manually destroy the object I'm exposing to Python when the process exits, and that seems to be doing the trick. No more driver error and core dump when the program exits :tada:. I don't really understand why the destructor was not getting run since I'm using `shared_ptr` to wrap the class I expose to python, but the fix doesn't seem so bad.
